src/main/java/io/netty/incubator/codec/http3/Http3FrameCodec.java

```diff
@@ -529,11 +614,112 @@ public void deregister(ChannelHandlerContext ctx, ChannelPromise promise) {
+        private static final int STATE_SUSPENDED = 0b1000_0000;
+        private static final int STATE_READ_PENDING = 0b0100_0000;
+        private static final int STATE_READ_COMPLETE_PENDING = 0b0010_0000;
+
+        private final ChannelHandlerContext ctx;
+        private final Http3FrameCodec codec;
+        private byte state;
+
+        ReadResumptionListener(ChannelHandlerContext ctx, Http3FrameCodec codec) {
+            this.ctx = ctx;
+            this.codec = codec;
+            assert codec.qpackAttributes != null;
+            if (!codec.qpackAttributes.dynamicTableDisabled() && !codec.qpackAttributes.decoderStreamAvailable()) {
+                codec.qpackAttributes.whenDecoderStreamAvailable(this);
+            }
+        }
+
+        void suspended() {
+            assert !codec.qpackAttributes.dynamicTableDisabled();
+            setState(STATE_SUSPENDED);
+        }
+
+        boolean readCompleted() {
+            if (hasState(STATE_SUSPENDED)) {
+                setState(STATE_READ_COMPLETE_PENDING);
+                return false;
+            }
+            return true;
+        }
+
+        boolean readRequested() {
+            if (hasState(STATE_SUSPENDED)) {
+                setState(STATE_READ_PENDING);
+                return false;
+            }
+            return true;
+        }
+
+        @Override
+        public void operationComplete(Future<? super QuicStreamChannel> future) {
+            if (future.isSuccess()) {
+                resume();
+            } else {
+                ctx.fireExceptionCaught(future.cause());
+            }
+        }
+
+        @Override
+        public void run() {
+            resume();
+        }
+
+        private void resume() {
+            unsetState(STATE_SUSPENDED);
+            try {
+                codec.channelRead(ctx, Unpooled.EMPTY_BUFFER);
+                if (hasState(STATE_READ_COMPLETE_PENDING)) {
+                    unsetState(STATE_READ_COMPLETE_PENDING);
+                    codec.channelReadComplete(ctx);
+                }
+                if (hasState(STATE_READ_PENDING)) {
+                    unsetState(STATE_READ_PENDING);
+                    codec.read(ctx);
+                }
+            } catch (Exception e) {
+                ctx.fireExceptionCaught(e);
+            }
+        }
+
+        private void setState(int toSet) {
+            state |= toSet;
+        }
+
+        private boolean hasState(int toCheck) {
+            return (state & toCheck) == toCheck;
+        }
+
+        private void unsetState(int toUnset) {
+            state &= ~toUnset;
+        }
+    }
+
+    /**
+     * A factory for {@link Http3FrameCodec}.
+     */
+    @FunctionalInterface
+    interface Http3FrameCodecFactory {
+        /**
+         * Creates a new {@link Http3FrameCodec} instance for the passed {@code streamType}.
+         *
+         * @param validator for the frames.
+         * @return {@link Http3FrameCodec}.
```

Comment:
nit: if you mention `Http3FrameCodec` here and above it also should return it as type. Maybe you just want to update the javadocs to say something like "returns a codec for Http3Frames" or something like this